Use global pubkey cache for block processing

paulhauner opened the pull request:
## Issue Addressed

- Resolves #848
- Related to #847

## Proposed Changes

- Uses the beacon chain pubkey cache for block signature verification.
- Changes `get_attesting_indices` to return a `Vec` instead of a `BTreeMap` to reduce heap allocations.
- Makes a (somewhat tangential) change to SSZ enc/dec for BLS objects to avoid unnecessary allocations.

## TODO:

- [x] Persist to disk to cover us in the case that the head has less validators than were previously seen.

## Additional Info

- ~~Blocked on #841~~

michaelsproul commented on eth2/state_processing/src/per_block_processing/block_signature_verifier.rs:
```diff
@@ -46,23 +47,34 @@ impl From<BlockOperationError<AttestationInvalid>> for Error {
+    F: Fn(usize) -> Option<Cow<'a, G1Point>> + Clone,
+{
     block: &'a SignedBeaconBlock<T>,
+    get_pubkey: Box<F>,
```
This compiles without the `Box`, do we need it?

paulhauner replied:
Removed, thank you!

paulhauner commented:
I addressed your comments, thank you! The `Option::filter` one was a great idea, I haven't used it before.

I've also commented out some tests failing due to `Eth2TestnetDir::hardcoded` introduced in #912 which broke when applied on top of #959.

paulhauner commented on beacon_node/rest_api/src/validator.rs:
```diff
@@ -189,11 +189,24 @@ fn return_validator_duties<T: BeaconChainTypes>(
+                .validator_index(&validator_pubkey)
+                .map_err(|e| {
+                ApiError::ServerError(format!("Unable to get validator index: {:?}", e))
+            })? {
```
Done! Great idea.
